### PR TITLE
[Merged by Bors] - feat: Adding/removing an element from a product of finsets

### PR DESCRIPTION
**Pull request by BoltonBailey**
Lemmas about the interaction of `Fintype.piFinset` with `Fin.consEquiv`, `Fin.snocEquiv`, `Fin.insertNthEquiv`.

---

- [x] depends on: #9028
- [x] depends on: #10545
- [x] depends on: #13568
- [x] depends on: #14665

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Fintype/Pi.lean [hunk 1]
```diff
@@ -133,3 +135,32 @@ lemma Fin.snoc_mem_piFinset_snoc_iff {n : ℕ} {α : Fin (n + 1) → Type*}
     (Sᵢ : (i : Fin n) → Finset (α i.castSucc)) (Sₙ : Finset (α <| .last n)) :
     Fin.snoc xs x ∈ Fintype.piFinset (Fin.snoc Sᵢ Sₙ) ↔ xs ∈ Fintype.piFinset Sᵢ ∧ x ∈ Sₙ := by
   simp_rw [Fin.mem_piFinset_succ_iff', init_snoc, snoc_last]
+
+lemma Finset.card_filter_piFinset_eq' {n : ℕ} {α : Fin (n + 1) → Type*}
+    (p : ((i : Fin n) → α i.succ) → Prop) [DecidablePred p]
+    (S : (i : Fin (n + 1)) → Finset (α i)) :
+    Finset.card (Finset.filter (fun r ↦ p (fun x ↦ r $ Fin.succ x)) (Fintype.piFinset S))
+      = Finset.card ((S 0) ×ˢ Finset.filter p (Fintype.piFinset (fun x ↦ S $ Fin.succ x))) := by
```
Comment (eric-wieser) [inline Mathlib/Data/Fintype/Pi.lean:143]: Instead of stating this as `card LHS = card RHS`, can you instead write `LHS = RHS.map _`, from which your result is trivial?

File: Mathlib/Data/Fintype/Pi.lean [hunk 2]
```diff
@@ -133,3 +135,26 @@ lemma Fin.snoc_mem_piFinset_snoc_iff {n : ℕ} {α : Fin (n + 1) → Type*}
...
   simp_rw [Fin.mem_piFinset_succ_iff', init_snoc, snoc_last]
+
+-- TODO Generalize from `succ` to `succAbove`. Requires `Fin.forall_fin_succAbove`.
+lemma Finset.map_piFinSuccAboveEquiv_filter_piFinset {n : ℕ} {α : Fin (n + 1) → Type*}
+    (p : ((i : Fin n) → α i.succ) → Prop) [DecidablePred p]
+    (S : (i : Fin (n + 1)) → Finset (α i)) :
+    Finset.map
+      ((Equiv.piFinSuccAboveEquiv α 0).toEmbedding)
+      (Finset.filter (fun r ↦ p (fun x ↦ r $ Fin.succ x)) (Fintype.piFinset S))
+      = ((S 0) ×ˢ Finset.filter p (Fintype.piFinset (fun x ↦ S $ Fin.succ x))) := by
```
Comment (eric-wieser) [inline Mathlib/Data/Fintype/Pi.lean:146]: Can you add a copy of this with `Equiv.symm` on the other side? There should be some lemma that lets you move the `map` across, so that you don't have to prove it again from scratch.

File: Mathlib/Data/Fintype/Pi.lean [hunk 3]
```diff
@@ -133,3 +135,26 @@ lemma Fin.snoc_mem_piFinset_snoc_iff {n : ℕ} {α : Fin (n + 1) → Type*}
...
+  ext ⟨x, f⟩
+  simp? [Fin.forall_fin_succ] says simp only [Fin.zero_succAbove, Fintype.mem_piFinset,
+      Fin.forall_fin_succ, and_imp, mem_map_equiv, mem_filter, mem_product]
+  tauto
+
+lemma Finset.card_filter_succ_piFinset_eq {n : ℕ} {α : Fin (n + 1) → Type*}
+    (p : ((i : Fin n) → α i.succ) → Prop) [DecidablePred p]
+    (S : (i : Fin (n + 1)) → Finset (α i)) :
+    Finset.card (Finset.filter (fun r ↦ p (fun x ↦ r $ Fin.succ x)) (Fintype.piFinset S))
+     = (S 0).card * Finset.card (Finset.filter p (Fintype.piFinset (fun x ↦ S $ Fin.succ x))) := by
```
Comment (eric-wieser) [inline Mathlib/Data/Fintype/Pi.lean:157]: Dot notation makes this more concise:
```suggestion
    ((Fintype.piFinset S).filter fun r ↦ p (fun x ↦ r <| Fin.succ x)).card
     = (S 0).card * ((Fintype.piFinset fun x ↦ S <| Fin.succ x).filter p).card := by
```

(also, `$` is a Lean3 -ism)

File: Mathlib/Data/Fintype/Pi.lean (hunk 2 above)
Comment (BoltonBailey) [reply Mathlib/Data/Fintype/Pi.lean]: Ok I added `Finset.filter_piFinset_eq_map_piFinSuccAboveEquiv_symm`, lmk if that's not what you were looking for.

File: Mathlib/Data/Fintype/Pi.lean [hunk 4]
```diff
@@ -133,3 +135,43 @@ lemma Fin.snoc_mem_piFinset_snoc_iff {n : ℕ} {α : Fin (n + 1) → Type*}
     (Sᵢ : (i : Fin n) → Finset (α i.castSucc)) (Sₙ : Finset (α <| .last n)) :
     Fin.snoc xs x ∈ Fintype.piFinset (Fin.snoc Sᵢ Sₙ) ↔ xs ∈ Fintype.piFinset Sᵢ ∧ x ∈ Sₙ := by
   simp_rw [Fin.mem_piFinset_succ_iff', init_snoc, snoc_last]
+
+lemma Finset.map_piFinSuccAboveEquiv_filter_piFinset_succAbove {n : ℕ} (k : Fin (n + 1))
+    {α : Fin (n + 1) → Type*}
+    (p : ((i : Fin n) → α (Fin.succAbove k i)) → Prop) [DecidablePred p]
+    (S : (i : Fin (n + 1)) → Finset (α i)) :
+    ((Fintype.piFinset S).filter fun r ↦ p (fun x ↦ r <| Fin.succAbove k x)).map
+      ((Equiv.piFinSuccAboveEquiv α k).toEmbedding)
```
Comment (YaelDillies) [inline Mathlib/Data/Fintype/Pi.lean:144]:
```suggestion
      (Equiv.piFinSuccAboveEquiv α k).toEmbedding
```
and below

File: Mathlib/Data/Fintype/Pi.lean [hunk 5]
```diff
@@ -133,3 +135,43 @@ lemma Fin.snoc_mem_piFinset_snoc_iff {n : ℕ} {α : Fin (n + 1) → Type*}
...
+    (p : ((i : Fin n) → α i.succ) → Prop) [DecidablePred p]
+    (S : (i : Fin (n + 1)) → Finset (α i)) :
+    ((Fintype.piFinset S).filter fun r ↦ p (fun x ↦ r <| Fin.succ x)).map
+      ((Equiv.piFinSuccAboveEquiv α 0).toEmbedding)
+    = S 0 ×ˢ (Fintype.piFinset (fun x ↦ S <| Fin.succ x)).filter p :=
+  Finset.map_piFinSuccAboveEquiv_filter_piFinset_succAbove 0 p S
+
+lemma Finset.filter_piFinset_eq_map_piFinSuccAboveEquiv_symm {n : ℕ} {α : Fin (n + 1) → Type*}
+    (p : ((i : Fin n) → α i.succ) → Prop) [DecidablePred p]
+    (S : (i : Fin (n + 1)) → Finset (α i)) :
```
Comment (YaelDillies) [inline Mathlib/Data/Fintype/Pi.lean:163]: I find this quite a lot more readable. Also use lowercase `s` to match the rest of the file.
```suggestion
    (S : ∀ i, Finset (α i)) :
```

**Review by YaelDillies — COMMENTED** [review]
Can you make sure you have the `Fin.castSucc`/`Fin.succ`/`Fin.succAbove` version of each lemma?

File: Mathlib/Data/Fintype/Pi.lean [hunk 6]
```diff
@@ -133,3 +135,43 @@ lemma Fin.snoc_mem_piFinset_snoc_iff {n : ℕ} {α : Fin (n + 1) → Type*}
...
+  ext ⟨x, f⟩
+  simp? [Fin.forall_iff_succAbove k] says simp only [Fintype.mem_piFinset,
+      Fin.forall_iff_succAbove k, and_imp, mem_map_equiv, Equiv.piFinSuccAboveEquiv_symm_apply,
+      mem_filter, Fin.insertNth_apply_same, Fin.insertNth_apply_succAbove, mem_product]
+  tauto
+
+lemma Finset.map_piFinSuccAboveEquiv_filter_piFinset {n : ℕ} {α : Fin (n + 1) → Type*}
+    (p : ((i : Fin n) → α i.succ) → Prop) [DecidablePred p]
+    (S : (i : Fin (n + 1)) → Finset (α i)) :
+    ((Fintype.piFinset S).filter fun r ↦ p (fun x ↦ r <| Fin.succ x)).map
```
Comment (YaelDillies) [inline Mathlib/Data/Fintype/Pi.lean:156]:
```suggestion
    ((Fintype.piFinset S).filter fun r ↦ p (Fin.tail r)).map
```
Note that the version for `Fin.castSucc` is `Fin.init` and the version for `Fin.succAbove` is non-existent?!

File: Mathlib/Data/Fintype/Pi.lean [hunk 7]
```diff
@@ -133,3 +135,43 @@ lemma Fin.snoc_mem_piFinset_snoc_iff {n : ℕ} {α : Fin (n + 1) → Type*}
     (Sᵢ : (i : Fin n) → Finset (α i.castSucc)) (Sₙ : Finset (α <| .last n)) :
     Fin.snoc xs x ∈ Fintype.piFinset (Fin.snoc Sᵢ Sₙ) ↔ xs ∈ Fintype.piFinset Sᵢ ∧ x ∈ Sₙ := by
   simp_rw [Fin.mem_piFinset_succ_iff', init_snoc, snoc_last]
+
+lemma Finset.map_piFinSuccAboveEquiv_filter_piFinset_succAbove {n : ℕ} (k : Fin (n + 1))
```
Comment (YaelDillies) [inline Mathlib/Data/Fintype/Pi.lean:139]: Those names are really hard to parse. Can you
* Refer to the unbundled function rather than the bundled `Equiv`. The fact that it's bundled is mostly an implementation detail stemming from using `Finset.map`.
* Use the correct `Equiv` rather than the general `Equiv.piFinSuccAboveEquiv`. We have `Equiv.piFinSucc`, but not `Equiv.piFinCastSucc`. So you first need a PR to add `Equiv.piFinCastSucc`, and it would be great if you also renamed `Equiv.piFinSuccAboveEquiv` to `Equiv.piFinSuccAbove` to match.

**Comment by BoltonBailey** [comment]
> Can you make sure you have the `Fin.castSucc`/`Fin.succ`/`Fin.succAbove` version of each lemma?

I got asked to add snoc-y API on #6605 as well by @eric-wieser. I have to say, I am not in favor of (me doing) this. I basically agree with Mac Malone here that [snoc is a code smell](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/List.2Econcat.20and.20List.2Eappend/near/406168233). Shouldn't we be trying to remove it from mathlib, rather than adding more of it?

**Comment by YaelDillies** [comment]
We're talking about `Fin.snoc`, which is used for proving, not `List.snoc`, which is used for programming (or maybe shouldn't be, as Mac says).

File: Mathlib/Data/Fintype/Pi.lean (hunk 5 above)
Comment (BoltonBailey) [reply Mathlib/Data/Fintype/Pi.lean]: Do you want me to do this everywhere a have a dependent function? Isn't it more stylish to use `∀` when and only when the return value is a Prop?

Comment (YaelDillies) [reply Mathlib/Data/Fintype/Pi.lean]: This new `(i : I) → ` notation is very unmathematical and verbose (I already `i` is an index, no need to spend an extra 15 characters to tell me it's a `Fin (n + 1)`).

**Comment by BoltonBailey** [comment]
I don't see why the difference matters, proofs should avoid being smelly too.

**Comment by YaelDillies** [comment]
? It's only smelly because it has unexpectedly bad complexity, but that's completely irrelevant for proving. I've personally needed both `Fin.cons` and `Fin.snoc` in the past and I will add those lemmas if you don't do it yourself.

**Comment by BoltonBailey** [comment]
The smell at play here for me is code duplication. As you have noted there is a whole parallel API for cons versions and snoc versions of things (init vs tail, succ vs castSucc). We are doing twice the work for an essentially equivalent set of theorems.

I'll look into where snoc is being used. My suspicion is that if we just wrote theorems the "right" way to start with, almost every theorem that is proved using snoc could be rephrased to be proved using cons, and judicious use of `reverse` could be used in places where it's really necessary to be dealing with both sides of the tuple.

**Comment by BoltonBailey** [comment]
At the very least, as I've suggested before, it should be handled by something like `@[to_additive]` so that we don't have to prove theorems twice.

File: Mathlib/Data/Fintype/Pi.lean (hunk 7 above)
Comment (BoltonBailey) [reply Mathlib/Data/Fintype/Pi.lean]: Ok the PRs renaming to remove the duplicated `Equiv` in the name and adding `Equiv.piFinCastSucc` are merged.

For the first point, I am looking at the definition of `Equiv.piFinSuccAbove` and not seeing `Finset.map`. I guess you are saying though saying that another def should be introduced for the function form of `Equiv.piFinSuccAbove` and `Equiv.piFinSuccAbove` should then be redefined as that?

Comment (BoltonBailey) [reply Mathlib/Data/Fintype/Pi.lean]: The inverse function is "insertNth". "extractNth" perhaps?

Comment (BoltonBailey) [reply Mathlib/Data/Fintype/Pi.lean]: #10545

Comment (YaelDillies) [reply Mathlib/Data/Fintype/Pi.lean]: That's a reasonable interpretation of my comment. `extractNth` seems like a reasonable name.

Comment (BoltonBailey) [reply Mathlib/Data/Fintype/Pi.lean]: Ok after doing all that, it seems like it's impossible to substitute this in because map actually takes an embedding, not a function. I am not sure if you wanted something else from me here?

File: Mathlib/Data/Fintype/Pi.lean [hunk 8]
```diff
@@ -134,3 +136,67 @@ theorem Finset.univ_pi_univ {α : Type*} {β : α → Type*} [DecidableEq α] [F
     (Finset.univ.pi fun a : α => (Finset.univ : Finset (β a))) = Finset.univ := by
   ext; simp
 #align finset.univ_pi_univ Finset.univ_pi_univ
+
+lemma Fin.mem_piFinset_succ_iff {n : ℕ} {α : Fin (n + 1) → Type*} (p : (i : Fin (n + 1)) → α i)
+    (S : (i : Fin (n + 1)) → Finset (α i)) :
```
Comment (YaelDillies) [inline Mathlib/Data/Fintype/Pi.lean:141]:
```suggestion
lemma Fin.mem_piFinset_succ_iff {n : ℕ} {α : Fin (n + 1) → Type*} {p : (i : Fin (n + 1)) → α i}
    {S : (i : Fin (n + 1)) → Finset (α i)} :
```

File: Mathlib/Data/Fintype/Pi.lean [hunk 9]
```diff
@@ -134,3 +136,67 @@ theorem Finset.univ_pi_univ {α : Type*} {β : α → Type*} [DecidableEq α] [F
...
+    p ∈ Fintype.piFinset S ↔ p 0 ∈ S 0 ∧ Fin.tail p ∈ Fintype.piFinset (Fin.tail S) := by
+  simp only [Fintype.mem_piFinset, forall_fin_succ, Fin.tail]
+
+lemma Fin.cons_mem_piFinset_cons_iff {n : ℕ} {α : Fin (n + 1) → Type*}
+    (x : α 0) (xs : (i : Fin n) → α i.succ)
+    (S₀ : Finset (α 0)) (Sᵢ : (i : Fin n) → Finset (α i.succ)) :
```
Comment (YaelDillies) [inline Mathlib/Data/Fintype/Pi.lean:147]:
```suggestion
    {x : α 0} {xs : (i : Fin n) → α i.succ}
    {S₀ : Finset (α 0)} {Sᵢ : (i : Fin n) → Finset (α i.succ)} :
```
etc

File: Mathlib/Data/Fintype/Pi.lean [hunk 10]
```diff
@@ -134,3 +136,67 @@ theorem Finset.univ_pi_univ {α : Type*} {β : α → Type*} [DecidableEq α] [F
...
+    Function.Embedding.equiv_toEmbedding_trans_symm_toEmbedding, map_refl]
+
+lemma Finset.card_filter_succ_piFinset_eq {n : ℕ} {α : Fin (n + 1) → Type*}
+    {p : ((i : Fin n) → α i.succ) → Prop} [DecidablePred p]
+    {S : (i : Fin (n + 1)) → Finset (α i)} :
+    ((Fintype.piFinset S).filter fun r ↦ p (Fin.tail r)).card
+    = (S 0).card * ((Fintype.piFinset (Fin.tail S)).filter p).card := by
+  rw [← Finset.card_map ((Equiv.piFinSuccAbove α 0).toEmbedding),
+    map_piFinSuccAbove_filter_piFinset]
+  exact Finset.card_product (S 0) ((Fintype.piFinset (Fin.tail S)).filter p)
```
Comment (YaelDillies) [inline Mathlib/Data/Fintype/Pi.lean:202]: You're missing lemmas here, right? There are three lemmas about `S 0`, one lemma about `S k` and no lemma about `S (Fin.last n)`.

Comment (BoltonBailey) [reply Mathlib/Data/Fintype/Pi.lean:202]: Well `Fin.mem_piFinset_succ_iff'` contains a reference to `S (Fin.last n)` on line 154. 

Here are the cons-snoc analogies as I see them
`Fin.mem_piFinset_succ_iff` is the cons version where  `Fin.mem_piFinset_succ_iff'` is the snoc version
`Fin.cons_mem_piFinset_cons_iff` is the cons version where `Fin.snoc_mem_piFinset_snoc_iff`
`Finset.map_piFinSuccAbove_filter_piFinset_succAbove` relates to an internal value `k`

`Finset.map_piFinSuccAbove_filter_piFinset` is `cons`-y, but trying the analogous `snoc`-y lemma, I get an error
```
lemma Finset.map_piFinSuccAbove_filter_piFinset_init {n : ℕ} {α : Fin (n + 1) → Type*}
    {p : ((i : Fin n) → α i.castSucc) → Prop} [DecidablePred p]
    {S : (i : Fin (n + 1)) → Finset (α i)} :
    ((Fintype.piFinset S).filter fun r ↦ p (Fin.init r)).map
      (Equiv.piFinSuccAbove α (Fin.last n)).toEmbedding
    = S (Fin.last n) ×ˢ (Fintype.piFinset (Fin.init S)).filter p -- error on this line
:= by 
  sorry
```
```
type mismatch
  S (Fin.last n) ×ˢ filter p (Fintype.piFinset (Fin.init S))
has type
  Finset (α (Fin.last n) × ((i : Fin n) → α (Fin.castSucc i))) : outParam (Type u_2)
but is expected to have type
  Finset (α (Fin.last n) × ((j : Fin n) → α (Fin.succAbove (Fin.last n) j))) : Type u_2
```
I am not sure why this poses a problem here while the `cons`-y one doesn't. Perhaps there is some definitional equality for that one. I suppose I could solve it with a cast but that seems like more complicated machinery. The last three lemmas all depend on this one so I can't prove the `snoc`-y version here.

Comment (YaelDillies) [reply Mathlib/Data/Fintype/Pi.lean]: That's because you're using `castSucc` in the finset but `succAbove (last n)` in the map. Those functions are equal but not defeq. Could you define `Equiv.piFinCastSucc` and use it here?

Comment (BoltonBailey) [reply Mathlib/Data/Fintype/Pi.lean]: [Equiv.piFinCastSucc](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/Equiv/Fin.html#Equiv.piFinCastSucc) already exists. It can't be used here because it isn't dependent the way [Equiv.piFinSuccAbove](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/Equiv/Fin.html#Equiv.piFinSuccAbove) is.

Comment (BoltonBailey) [reply Mathlib/Data/Fintype/Pi.lean]: In fact, I see that this very [Equiv.piFinCastSucc](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/Equiv/Fin.html#Equiv.piFinCastSucc) function was something [you asked me to add](https://github.com/leanprover-community/mathlib4/pull/7898#discussion_r1423557280) for this PR back in December to match [Equiv.piFinSucc](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/Equiv/Fin.html#Equiv.piFinSucc).

Comment (BoltonBailey) [reply Mathlib/Data/Fintype/Pi.lean]: When I try to make is dependent, I get a similar error which again doesn't happen for the cons-y version
```
def Equiv.piFinCastSucc (n : ℕ) (α : Fin (n + 1) → Type u) : 
   ((i : Fin (n + 1)) → α i) ≃ (α (.last _)) × ((j : Fin n) → α (.castSucc j)) :=
  Equiv.piFinSuccAbove α (.last _)
```
```
type mismatch
  piFinSuccAbove α (Fin.last n)
has type
  ((j : Fin (n + 1)) → α j) ≃ α (Fin.last n) × ((j : Fin n) → α (Fin.succAbove (Fin.last n) j)) : Type u
but is expected to have type
  ((i : Fin (n + 1)) → α i) ≃ α (Fin.last n) × ((j : Fin n) → α (Fin.castSucc j)) : Type u
```

Comment (YaelDillies) [reply Mathlib/Data/Fintype/Pi.lean]: I mean yeah you can't use `piFinSuccAbove` to define it (at least without some `cast` somewhere).

I agree that making `Equiv.piFinCastSucc` dependent is the correct move.

Comment (BoltonBailey) [reply Mathlib/Data/Fintype/Pi.lean]: To be clear, I am not trying to imply I do or don't think making `Equiv.piFinCastSucc` dependent is the correct move. I am just trying to point out that there are difficulties in doing so. 

I would like to express again that trying to develop `snoc` related API is not what I would like to be doing with my time. I think that probably this work would be better off being done by someone who has a better sense of why it's important, and therefore a better sense of the design considerations involved.

Comment (YaelDillies) [reply Mathlib/Data/Fintype/Pi.lean]: > I would like to express again that trying to develop `snoc` related API is not what I would like to be doing with my time. I think that probably this work would be better off being done by someone who has a better sense of why it's important, and therefore a better sense of the design considerations involved.

The problem is that, by doing so, you are making the parallel APIs diverge, thus decreasing the quality of mathlib as a whole.

I have now opened #13567 and #13568 to add the missing prerequisites. I will probably need one further PR before yours can be merged.

**Comment by YaelDillies** [comment]
Sorry for the long delay. Everything should be good now.

maintainer merge

File: Mathlib/Data/Fintype/Pi.lean [hunk 11]
```diff
@@ -174,3 +175,94 @@ lemma piDiag_subset_piFinset [DecidableEq ι] [Fintype ι] :
     s.piDiag ι ⊆ Fintype.piFinset fun _ ↦ s := by simp [← piFinset_filter_const]
 
 end Finset
+
```
Comment (eric-wieser) [inline Mathlib/Data/Fintype/Pi.lean:178]: Could you add a section heading (`/-! ### title -/`) here? Perhaps "Results about finsets of tuples" or similar?

File: Mathlib/Data/Fin/Tuple/Finset.lean
Comment (eric-wieser) [inline Mathlib/Data/Fin/Tuple/Finset.lean:1]: @YaelDillies, you split this file from `Data.Fintype.Pi` in #9383; are you happy with it being put back? Or is this a bad merge conflict?

Comment (YaelDillies) [reply Mathlib/Data/Fin/Tuple/Finset.lean:1]: Hmm, I guess I am going back and forth on whether this file should be split or not. For now I have for keeping this split

**Comment by jcommelin** [comment]
Removing `maintainer-merge`, since it has been 2 weeks, and CI is failing.

**Comment by YaelDillies** [comment]
Should be good now.

maintainer merge

**Review by jcommelin — COMMENTED** [review]
bors d=YaelDillies

**Comment by YaelDillies** [comment]
bors merge

**Comment by YaelDillies** [comment]
bors merge
